File: Chapter.13.cpp

class Breed
{
private:
    int health_;
    const char *attack_;

public:
    Breed(int health, const char *attack) : health_(health), attack_(attack) {}
    int getHealth() { return health_; }
    const char *getAttack() { return attack_; }
    Monster *newMonster()
    {
        return new Monster(*this);
    }
};

class Monster
{
    friend class Breed; // Breed 成为 Monster 的友元从而能够访问 Monster 的构造方法。

private:
    int health_;
    Breed &breed_;

private:
    Monster(Breed &breed) : health_(breed.getHealth()), breed_(breed) {}
};

// 修改前的创建怪物方法
// Monster *monster = new Monster(someBreed);

// 修改后的代码：
// Monster *monster = someBreed.newMonster()；

File: Chapter.13.py
class Monster:
    def __init__(self, breed: Breed):
        self._health = breed.getHealth()
        self._breed = breed

    def getAttack(self):
        if self._health < LOW_HEALTH:
            return "The monster fails weakly."

    def getBreed(self):
        "Breed 本质上一个类型，而 Monster 是无类型的——需要实例化才能使用。"
        return self._breed


class Breed:
    def __init__(
        self, health: int = None, attack: function = None, parent: Breed = None
    ):
        if parent is not None:
            if health == 0:
                self._health = parent.getHealth()
            if attack is None:
                self._attack = parent.getAttack()

    def newMonster(self):
        """
        产生新 Moster 实例的工厂方法，使用“类型”来产生一个类型
        """
        return Monster(self)

    def getHealth(self):
        return self._health

    def getAttack(self):
        return self._attack

